pkgs/applications/graphics/structorizer/default.nix

```
@@ -0,0 +1,105 @@
+{ stdenv
+, lib
+, fetchFromGitHub
+, jdk11
+, makeDesktopItem
+, makeWrapper
+, copyDesktopItems
+}:
+
+stdenv.mkDerivation rec {
+  pname = "structorizer";
+  version = "3.32-11";
+
+  desktopItems = [
+    (makeDesktopItem {
+      type = "Application";
+      name = "Structorizer";
+      desktopName = "Structorizer";
+      genericName = "Diagram creator";
+      comment = meta.description;
+      icon = pname;
+      exec = pname;
+      terminal = false;
+      mimeTypes = [ "application/nsd" ];
+      categories = [
+        "Development"
+        "Graphics"
+        "VectorGraphics"
+        "RasterGraphics"
+        "ComputerScience"
+      ];
+      keywords = [ "nsd" "diagrams" ];
+    })
+  ];
+
+  src = fetchFromGitHub {
+    owner = "fesch";
+    repo = "Structorizer.Desktop";
+    rev = version;
+    hash = "sha256-rGyeOcGm6uBplgTjMIOy/xRekfHacwDy9kkMigmRSdk=";
+  };
+
+  nativeBuildInputs = [ jdk11 makeWrapper copyDesktopItems ];
+
+  patches = [ ./makeStructorizer.patch ./makeBigJar.patch ];
+
+  postPatch = ''
+    chmod +x makeStructorizer
+    chmod +x makeBigJar
+
+    patchShebangs --build makeStructorizer
+    patchShebangs --build makeBigJar
+  '';
+
+  buildPhase = ''
+    runHook preBuild
+
+    ./makeStructorizer
+    ./makeBigJar
+
+    runHook postBuild
+  '';
+
+  installPhase = ''
+    runHook preInstall
+
+    install -D ${pname}.jar -t $out/share/java/
+    install -d $out/bin
+    makeWrapper ${jdk11}/bin/java $out/bin/${pname} \
+      --add-flags "-jar $out/share/java/${pname}.jar"
+
+    mkdir -p $out/share/mime/application
+    cat << EOF > $out/share/mime/application/nsd.xml
```

```suggestion
    cat << EOF > $out/share/mime/packages/structorizer.xml
```
I found out that apperently, mime-types provided by packages are supposed to go under mime/packages. Based on what I've seen by other packages so far, I'd suggest to name the file `structorizer.xml`, but keep the mimetype as application/nsd.

I tested it by commiting the new path, rebasing onto upstream/nixos-unstable and then temorarily setting the nixpkgs input in my `flake.nix` to my local nixpkgs clone. Maybe there is a cleaner way, but idk.

It correctly installed the xml-file to `/run/current-system/sw/share/mime/packages/nsd.xml`, but didn't work right away. I ran `update-mime-database /run/current-system/sw/share/mime` and it gave me some error-messages, which I fixed by modifiying the xml file. Now i got it to work.

---

> I tested it by commiting the new path, rebasing onto upstream/nixos-unstable and then temorarily setting the nixpkgs input in my flake.nix to my local nixpkgs clone. Maybe there is a cleaner way, but idk.

What about `nix-env -f . -iA structorizer`?

---

How do I test tho, if it works?

---

> I ran update-mime-database /run/current-system/sw/share/mime and it gave me some error-messages, which I fixed by modifiying the xml file. Now i got it to work.

I can't update the db or edit the XML file:
```bash
> update-mime-database /run/current-system/sw/share/mime
I/O error : Read-only file system
I/O error : Read-only file system
Failed to write XML file; For permission problems, try rerunning as root
```
```bash
> chmod +w -R /run/current-system/sw/share/mime
…
chmod: changing permissions of '/run/current-system/sw/share/mime/version': Read-only file system
```

---

> How do I test tho, if it works?

I just did `gio info file.nsd`.

---

> I can't update the db or edit the XML file:

Okay, maybe I wasn't concise enough, so i'll just reelaborate:

On a normal linux system, a package would install it's xml file under `/usr/share/mime` and then you'd have to run `update-mime-database`. That will rebuild the mime cache by rewriting the binary file `/usr/share/mime/mime.cache`, and as a neat bonus, emit some error-messages if some of the xml files aren't valid. On NixOS however, `/run/current-system/sw/share/mime` is read-only, so you can't rebuild the mime cache or edit the xml-files directly. When you rebuild NixOS, it'll symlink together all your packages and then rebuild the mime cache for you.

When I said "modify the xml file", I meant: "modify the heredoc that is written to the xml file". And once you've done and your xml is valid, you won't have to run `update-mime-database`, NixOS will do that for you on rebuild. I just used it for debugging.

---

> What about `nix-env -f . -iA structorizer`?

apparently, nix-env doesn't build a mime.cache in `~/.nix-profile/share/mime`. I don't know how to create one, since the nix-profile is read-only.

---

> I tested it by commiting the new path, rebasing onto upstream/nixos-unstable and then temorarily setting the nixpkgs input in my flake.nix to my local nixpkgs clone. Maybe there is a cleaner way, but idk.

Ok, so I did that too and it works now. I am not content with the mime type not working when Structorizer is installed in the nix profile.

---

No, wait… `~/.nix-profile/share/mime/mime.cache` does exist on my system. Does Home Manager create it? Then it'd at least work with Home Manager.

---

>Does Home Manager create it?

That's possible, since i don't use home-manager.

---

Yeah, 't works. Just tested it.

---

I asked in the Nix matrix channel if it is possible to update the mime cache with `nix-env`.

---

None replied.